ddl/multi_schema_change.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl
+
+import (
+	"sync"
+
+	"github.com/pingcap/tidb/meta"
+	"github.com/pingcap/tidb/parser/model"
+)
+
+func onMultiSchemaChange(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	if job.MultiSchemaInfo.Revertible {
+		// Handle the rolling back job.
+		if job.IsRollingback() || job.IsCancelling() {
+			// Rollback/cancel the sub-jobs in reverse order.
+			for i := len(job.MultiSchemaInfo.SubJobs) - 1; i >= 0; i-- {
+				sub := job.MultiSchemaInfo.SubJobs[i]
+				if isFinished(sub) {
+					continue
+				}
+				proxyJob := cloneFromSubJob(job, sub)
+				ver, err = w.runDDLJob(d, t, proxyJob)
+				mergeBackToSubJob(proxyJob, sub)
+				if i == 0 {
+					// The last rollback/cancelling sub-job is done.
+					if job.IsRollingback() {
+						job.State = model.JobStateRollbackDone
+					}
+					if job.IsCancelling() {
+						job.State = model.JobStateCancelled
+					}
+				}
+				return ver, err
+			}
+		}
+
+		// The sub-jobs are normally running.
+		// Run the first executable sub-job.
+		for i, sub := range job.MultiSchemaInfo.SubJobs {
+			if !sub.Revertible {
+				// Skip the sub jobs which related schema states
+				// are in the last revertible point.
+				continue
+			}
+			proxyJob := cloneFromSubJob(job, sub)
+			ver, err = w.runDDLJob(d, t, proxyJob)
+			mergeBackToSubJob(proxyJob, sub)
+			handleRevertibleException(job, sub.State, i)
+			return ver, err
+		}
+		// All the sub-jobs are non-revertible.
+		job.MultiSchemaInfo.Revertible = false
+		// Step the sub-jobs to the non-revertible states all at once.
+		for _, sub := range job.MultiSchemaInfo.SubJobs {
```

Comment:
why need this here? Maybe L74-L83 is enough.

Reply:
This is used to step all the schema object to a "visible" state. Note that there is no "return".

Take `add column a, add column b` as an example, this is used to change `a` and `b` to public in a single transaction.